### PR TITLE
fix(auth): make sudo to be aware of custom auth backend

### LINKED ISSUE #10380: Confirming password impossible with custom authentication backend
## Important Details

How are you running Sentry?

* [x] On-Premise docker [Version 9.0]
* [ ] Saas (sentry.io)
* [ ] Other [briefly describe your environment]

## Description

I have a custom authentication backend which use my corporate authentication provider. The backend will create a new user on first login. The company policy doesn't allow to store user passwords in application databases, so the password will be empty in Sentry DB. In future if the user would like to log in, the password will be validated with corporate authentication again. This functionality works fine by registering custom authentication backend functionality in Django via sentry.conf.py file.  
The problem is when the user wants to perform an admin action. Sentry wants the user to confirm password. To do that, it uses `check_password()` method from the `User` model. The problem is that `check_password()` is testing the password against the one stored in a database. If user has logged in via custom backend which doesn't store the password, it is impossible to confirm the password.  

## Steps to Reproduce

1. Log in with custom backend which doesn't store password in database
2. Try to perform any administrative action e.g. removing a project
3. Your password will be rejected every time.

### What you expected to happen

Sentry should use the custom backend to validate the user password.

### Possible Solution

This can be fixed by overriding or monkey-patching the `check_password()` method in the User model, but it doesn't seem to be the clean solution. Especially when you're using multiple custom backends (we do that in our company).   

I think, the `check_password()` usage shall be replaced with `authenticate()` function which is defined by backend. This would be the easiest and cleanest solution.

### PR DESCRIPTION
**Pull request by aklajnert**
The current behavior won't change, except that it will be possible to confirm your password in admin panel if custom authentication backend doesn't store password in DB.

Fixes GH-10380

### CODE REVIEW

**Comment by markstory** [comment]
How does this behave if the user is using an Sso provider like github or google?

**Comment by aklajnert** [comment]
I have no idea how to setup the SSO in dev environment. I can't find any docs about it.
Any dev docs about how to add a provider and set it up? I'm trying to figure it out from the code, but no luck...

**Review by BYK — APPROVED** [review]
Okay, finally tested this. It was quite tricky as setting up the GitHub integration requires and organization and then a member in the org that has a different email than your initial super user (otherwise this still prompts for a password).

It seems to work and keeps the old behavior so approving.

**Comment by BYK** [comment]
Asking @getsentry/security for a final approval before merging.

**Review by leedongwei — APPROVED** [review]
👀 

I didn't test the flow, but the code diff looks reasonable.

**Review by dcramer — COMMENTED** [review]
is it possible this is something we could add a test for? just prefer the rigor around auth code to ensure we're not causing any issues (and that if we choose to support something we dont break it accidentally in the future)

File: src/sentry/api/endpoints/auth_index.py [hunk 1]
```diff
@@ -115,7 +115,12 @@ def put(self, request):
 
         # attempt password authentication
         else:
-            authenticated = request.user.check_password(validator.validated_data["password"])
+            authenticated = (
+                authenticate(
+                    username=request.user.username, password=validator.validated_data["password"]
+                )
+                is not None
```
Comment (aklajnert) [inline src/sentry/api/endpoints/auth_index.py:122]: @BYK if the `authenticate()` function will return `False`, value of `authenticated` will be `True`. I don't think it is the right behavior here.

Comment (BYK) [reply src/sentry/api/endpoints/auth_index.py:122]: What I'm told by @mattrobenolt was `authenticate` returning a `User` model or `None` so this should be okay. The docs also show a very similar use here: https://docs.djangoproject.com/en/1.11/topics/auth/default/#authenticating-users

Comment (aklajnert) [reply src/sentry/api/endpoints/auth_index.py:122]: Oh, I was thinking that this function can return `False`, but you're right.

**Comment by aklajnert** [comment]
Whoops, I've missed the bot warning. @BYK - I'm not sure what happens now - move on or reopen this?  
No pressure, I have a workaround since the beginning, but might be useful for others.

**Comment by BYK** [comment]
@aklajnert sorry about the closure, I missed the bot's comment too. I've been trying to get around to testing this one more time and get it merged.

Very sorry for the long delays here :(

**Pull request closed without merging**